Setup default session store internally, no longer through an initializer

prathamesh-sonpatki opened the pull request:
- By default the session store will be set to cookie store with application name as session key.
- Based on comment by DHH here - https://github.com/rails/rails/issues/25181#issuecomment-222312764.

kamipo commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -67,5 +67,10 @@ class Railtie < Rails::Railtie #:nodoc:
+
+    initializer "setup session store" do |app|
+      app_name = app.railtie_name.sub(/_application$/, '')
+      app.config.session_store :cookie_store, key: "'_#{app_name}_session'"
```
Should be `"_#{app_name}_session"`?

prathamesh-sonpatki replied:
Fixed, thanks

rafaelfranca commented:
Tests are broken.

prathamesh-sonpatki commented:
@rafaelfranca Please review. Build is green.

rafaelfranca commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -67,5 +67,12 @@ class Railtie < Rails::Railtie #:nodoc:
     end
+
+    initializer "setup session store" do |app|
+      ActiveSupport.on_load(:action_controller) do
```
Why does it need to be inside on_load hook if it sets a config?

rafaelfranca replied:
This config lives in the railties so the initializer should be there too, not in Action Controller.

prathamesh-sonpatki replied:
We have lot of tests which set the session_store in `config/application.rb`. If this is not in the inside of on_load hook, then it gets run after the application.rb is run -- overriding the config set in application.rb.

rafaelfranca replied:
This is just masking a problem. It only works because ActionController::Base is not loaded in those tests.

prathamesh-sonpatki replied:
> This config lives in the railties so the initializer should be there too, not in Action Controller.

I was actually confused here, I thought of railties also but where do we place it in railties?

prathamesh-sonpatki replied:
@rafaelfranca We can move it to Rails::Application::Configuration class. What do you think?

rafaelfranca replied:
I mean `railties` folder. This configuration have nothing to do with Action Controller. All work is being done in `Application::Configuration` class so this initializers should be in the `application.rb` file.

rafaelfranca replied:
`Rails::Application::Configuration` doesn't define any initializer. It should be in bootstrap or finisher. I believe finisher is the right place to put since we only want to set it if it was not already set.

kaspth commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -67,5 +67,12 @@ class Railtie < Rails::Railtie #:nodoc:
+
+    initializer "setup session store" do |app|
+      ActiveSupport.on_load(:action_controller) do
+        app_name = app.class.name ? app.railtie_name.sub(/_application$/, '') : ''
```
Could make good use of `chomp`: `app.railtie_name.chomp('_application')`.

Also when won't the app class have a `name`? Is it likely that `Rails::Application` subclasses are an anonymous class?

prathamesh-sonpatki replied:
For few tests I noticed that the app_name was coming out to be empty. I will revert this locally and post which were those tests.

prathamesh-sonpatki replied:
@kaspth The problem was with all the tests using `make_basic_app` helper because of this line - https://github.com/prathamesh-sonpatki/rails/blob/f8244070e8098d930377d3ef99bf3d44fa7d4e83/railties/test/isolation/abstract_unit.rb#L172

kaspth commented on guides/source/configuring.md:
```diff
@@ -142,7 +142,7 @@ defaults to `:debug` for all environments. The available log levels are: `:debug
 * `config.public_file_server.enabled` configures Rails to serve static files from the public directory. This option defaults to `true`, but in the production environment it is set to `false` because the server software (e.g. NGINX or Apache) used to run the application should serve static files instead. If you are running or testing your app in production mode using WEBrick (it is not recommended ...
 
-* `config.session_store` is usually set up in `config/initializers/session_store.rb` and specifies what class to use to store the session. Possible values are `:cookie_store` which is the default, `:mem_cache_store`, and `:disabled`. The last one tells Rails not to deal with sessions. Custom session stores can also be specified:
+* `config.session_store` specifies what class to use to store the session. Possible values are `:cookie_store` which is the default, `:mem_cache_store`, and `:disabled`. The last one tells Rails not to deal with sessions. Defaults to cookie store with application name as session key. Custom session stores can also be specified:
```
`Defaults to a cookie store with the application name as the session key` 😁

kaspth commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -170,20 +170,23 @@ def session_store(*args)
         else
           @session_store = args.shift
+
+          if @session_store == :active_record_store
```
Curious, why do we have to move this?

rafaelfranca replied:
Why this was moved?

prathamesh-sonpatki replied:
Please check https://github.com/rails/rails/pull/25438/commits/c73107e83e8f6fe420941abebe3521bfc05e2b9a :smile:

kaspth replied:
Faster than @rafaelfranca! Woo \o/

<img width="498" alt="skaermbillede 2016-07-06 kl 19 03 28" src="https://cloud.githubusercontent.com/assets/350807/16627177/698f2244-43ac-11e6-9e04-fc99dfb4078a.PNG">

prathamesh-sonpatki replied:
I have another proposal of changing the `config.session_store` to not act as accessor as well as setter at the same time, but I was planning to propose it after this PR :)

kaspth replied:
Does that benefit outweigh a deprecation cycle? Seems a bit to quick in my opinion.

I don't quite understand your reasoning behind the move. I'm curious which tests were failing?

prathamesh-sonpatki replied:
> Does that benefit outweigh a deprecation cycle? Seems a bit to quick in my opinion.

Yup, it does. 

> I'm curious which tests were failing?

With my earlier implementation, the test https://github.com/rails/rails/blob/81b6bc99754e33f31d4ed61306680ae2c6f68045/railties/test/application/configuration_test.rb#L1181 was failing. With the current implementation, it no longer fails but I have kept the commit as it is. Because I feel the error about missing `activerecord-session_store` gem should be shown when someone tries to set `config.session_store` to `:activerecord_store` and not when somebody tries to access `config.session_store`.

Currently the same `session_store` method is used for accessing as well as setting the session store.

``` ruby
def session_store(*args)
        if args.empty?
          # Access session store
          case @session_store
          when :disabled
            nil
          when :active_record_store
            begin
              ActionDispatch::Session::ActiveRecordStore
            rescue NameError
              raise "`ActiveRecord::SessionStore` is extracted out of Rails into a gem. " \
                "Please add `activerecord-session_store` to your Gemfile to use it."
            end
          when Symbol
            ActionDispatch::Session.const_get(@session_store.to_s.camelize)
          else
            @session_store
          end
        else
          # Set session store
          @session_store = args.shift
          @session_options = args.shift || {}
        end
      end
```
This is the implementation from master branch which raises the error in the code path of `accessing session store`. This change moves the error from `accessing session store` to `setting session store` code path. The behavior is same. IMO it will be good if we raise as early as possible.

prathamesh-sonpatki commented:
@rafaelfranca Please review.

prathamesh-sonpatki commented on railties/lib/rails/application/finisher.rb:
```diff
@@ -33,6 +33,14 @@ module Finisher
 
+      # Setup default session store if not already set in config/application.rb
+      initializer :setup_default_session_store, before: :build_middleware_stack do |app|
+        unless app.config.session_store
```
This introduces one issue. If user has set the `session_store` to `:disabled` then this change overwrites it.

kaspth replied:
Why not just add `app.config.session_store && app.config.session_store == :disabled`? 😁

prathamesh-sonpatki replied:
@kaspth Because `app.config.session_store ` returns `nil` when it is set to `disabled`. 

There is no way as of now to see if the session store is not set without touching internal instance variables.

kaspth commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -170,20 +168,23 @@ def session_store(*args)
           case @session_store
```
Could also shift to kw args while we're here:

```ruby
def session_store(new_session_store = nil, **options)
  unless new_session_store
    case @session_store
    # ...
  else
    @session_store   = new_session_store
    @session_options = options
  end
end
```

prathamesh-sonpatki committed:
Introduce a new method for checking whether session_store is set by user or by Rails

- We need to ability to check whether the session store it is set or
   not so that we can set it ourselves in an internal initializer.
- We can't rely on return value of `config.session_store` as it can
  return `nil` when set to `disabled` and we will accidentally override it
  and set to default cookie store.
- So introduced new method which just tells us whether it is set or not.

prathamesh-sonpatki committed:
Refactor the existing session_store to use keyword args and raise early when set to activerecord session store

- Use keyword args as it is possible to use them now.
- The error message for activerecord-session_store gem was added in 1807384.
- But it was added for a code path which gets called when we try to
  **access** the session store, not when we **set** it.
- So the test expecting the exception started failing because now the
  session store is set via railtie again **after** setting it first with
  :active_record_store in the test.
- As the error is not raised while setting the store to
  :active_record_store, the store gets overwritten by railtie and when
  we access it via `session_store` while building the default middleware
  stack, the exception is not raised.
- This commit moves the code for raising the exception to the path where
  we try to set the store.

prathamesh-sonpatki committed:
Setup default session store internally, no longer through an application initializer

- By default the session store will be set to cookie store with
  application name as session key.
- Older apps are not affected as they will have the session store
  initializer generated by Rails in older versions, and Rails will not
  overwrite the session store if it is already set or disabled.
- But new apps will not have the initializer, instead the session store
  will be set to cookie store by default.
- Based on comment by DHH here - https://github.com/rails/rails/issues/25181#issuecomment-222312764.

prathamesh-sonpatki commented:
@kaspth @rafaelfranca Ready for review.

rafaelfranca commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -165,29 +163,38 @@ def colorize_logging=(val)
-          when :active_record_store
+      def session_store(new_session_store = nil, **options)
+        if new_session_store
+
```
✂️

rafaelfranca commented:
Just a really small comment, but this really looks good 👏

rafaelfranca commented:
Merged
